File: ml/availability_facts.py
"""Every figure the availability page publishes, as a machine-readable artifact.

Writes ``ml/availability_facts.json``, which ``src/lib/availability-facts.ts`` mirrors and
``src/lib/__tests__/availability-facts.test.ts`` pins. A number edited in a component and
nowhere else fails the suite — the same arrangement the Playoff Rest page uses, and for the
same reason: its figures were once retyped into three places and drifted.

The measurement itself is explained in ``ml/availability_cost.py`` and
``ml/availability_quality.py``. This file re-runs the parts that get published, at one
specification, and records the sample it came from.

**Everything here is retrospective.** Who sat is known only because the game was played, so
none of it forecasts anything, and the page has to say so.
"""

from __future__ import annotations

import json
import logging
from collections import defaultdict, deque
from pathlib import Path

import numpy as np
import pandas as pd

logging.basicConfig(level=logging.INFO, format="%(message)s")
log = logging.getLogger(__name__)

ROOT = Path(__file__).resolve().parent.parent
DATA = ROOT / "ml" / "data"
SHOOTING = DATA / "shooting"
FEATURES = DATA / "fatigue_features.csv"
TABLE = DATA / "fatigue_model_table.csv"
JSON_PATH = ROOT / "ml" / "availability_facts.json"

WINDOW = 5
ROTATION_MINUTES = 15.0
MIN_APPEARANCES = 2
QUALITY_GAMES = 20
QUALITY_MIN_GAMES = 5

SCHEDULE = ["d_is_b2b", "d_alt_visit", "d_prior_ot", "d_density_points"]
CONTROLS = ["d_strength", "d_season_day"]


def parse_minutes(value: object) -> float:
    if not isinstance(value, str) or ":" not in value:
        return 0.0
    mm, _, ss = value.partition(":")
    try:
        return int(mm) + int(ss) / 60.0
    except ValueError:
        return 0.0


def ols(x: np.ndarray, y: np.ndarray, names: list[str]) -> tuple[pd.DataFrame, float]:
    z = np.column_stack([np.ones(len(x)), x])
    beta, *_ = np.linalg.lstsq(z, y, rcond=None)
    resid = y - z @ beta
    sigma2 = float(resid @ resid) / (len(y) - z.shape[1])
    se = np.sqrt(np.diag(sigma2 * np.linalg.pinv(z.T @ z)))
    frame = pd.DataFrame(
        {"feature": ["intercept", *names], "beta": beta, "se": se, "t": beta / se}
    ).set_index("feature")
    return frame, float(np.sqrt(np.mean(resid**2)))


def build_team_games() -> pd.DataFrame:
    box = pd.concat(
        [pd.read_csv(p, low_memory=False) for p in sorted(SHOOTING.glob("player_boxscores_*.csv"))],
        ignore_index=True,
    )
    box["min"] = box["minutes"].map(parse_minutes)
    box["external_id"] = box["game_id"].astype("int64").astype(str).str.zfill(10)

    z = box.fillna(
        {
            c: 0
            for c in [
                "points", "field_goals_made", "field_goals_attempted", "free_throws_made",
                "free_throws_attempted", "rebounds_offensive", "rebounds_defensive",
                "steals", "assists", "blocks", "fouls_personal", "turnovers",
            ]
        }
    )
    box["game_score"] = (
        z["points"]
        + 0.4 * z["field_goals_made"]
        - 0.7 * z["field_goals_attempted"]
        - 0.4 * (z["free_throws_attempted"] - z["free_throws_made"])
        + 0.7 * z["rebounds_offensive"]
        + 0.3 * z["rebounds_defensive"]
        + z["steals"]
        + 0.7 * z["assists"]
        + 0.7 * z["blocks"]
        - 0.4 * z["fouls_personal"]
        - z["turnovers"]
    )

    sides = pd.read_csv(
        FEATURES,
        usecols=["game_id", "external_id", "date", "season", "team_id", "is_home", "status", "game_type"],
        low_memory=False,
    )
    sides = sides[(sides["status"] == "final") & (sides["game_type"] == "regular")].copy()
    sides["external_id"] = sides["external_id"].astype(str).str.zfill(10)
    sides["side"] = np.where(sides["is_home"] == 1, "home", "away")

    # hoopR carries its own team_id, game_id and season, all of which collide with ours and
    # would silently become _x / _y. The join key is (external_id, side); none are needed.
    play = box.drop(columns=["team_id", "game_id", "season"]).merge(
        sides[["game_id", "external_id", "date", "season", "team_id", "side"]],
        on=["external_id", "side"],
        how="inner",
    )

    appeared = play[play["min"] > 0].sort_values(["person_id", "date"]).copy()
    appeared["q_gs"] = appeared.groupby("person_id")["game_score"].transform(
        lambda s: s.shift(1).rolling(QUALITY_GAMES, min_periods=QUALITY_MIN_GAMES).mean()
    )
    play = play.merge(
        appeared[["game_id", "person_id", "q_gs"]], on=["game_id", "person_id"], how="left"
    ).sort_values(["team_id", "date", "person_id"])

    history: dict[int, deque] = defaultdict(lambda: deque(maxlen=WINDOW))
    rows = []
    for (team_id, game_id), grp in play.groupby(["team_id", "game_id"], sort=False):
        window = history[team_id]
        played_now = dict(zip(grp["person_id"], grp["min"]))

        mins: dict[int, list] = defaultdict(list)
        qual: dict[int, float] = {}
        for past in window:
            for pid, rec in past.items():
                if rec["min"] > 0:
                    mins[pid].append(rec["min"])
                    if not np.isnan(rec["q_gs"]):
                        qual[pid] = rec["q_gs"]

        rotation = {
            pid: float(np.mean(v))
            for pid, v in mins.items()
            if len(v) >= MIN_APPEARANCES and float(np.mean(v)) >= ROTATION_MINUTES
        }

        if len(window) == WINDOW and rotation:
            known = [qual[p] for p in rotation if p in qual]
            replacement = float(np.median(known)) if known else 0.0
            best = max(rotation, key=lambda p: qual.get(p, -np.inf)) if known else None

            miss_value = 0.0
            miss_count = 0
            best_out = 0
            for pid, _avg in rotation.items():
                if played_now.get(pid, 0.0) > 0.0:
                    continue
                miss_count += 1
                if pid in qual:
                    miss_value += max(0.0, qual[pid] - replacement)
                    if pid == best:
                        best_out = 1
            rows.append(
                {
                    "game_id": game_id,
                    "team_id": team_id,
                    "season": grp["season"].iloc[0],
                    "miss_value": miss_value,
                    "miss_count": miss_count,
                    "best_out": best_out,
                    "rotation_size": len(rotation),
                }
            )

        window.append(
            {
                pid: {"min": m, "q_gs": q}
                for pid, m, q in zip(grp["person_id"], grp["min"], grp["q_gs"])
            }
        )

    return pd.DataFrame(rows)


def main() -> None:
    log.info("building team-game availability")
    av = build_team_games()
    log.info("team-games measured: %d", len(av))

    sides = pd.read_csv(FEATURES, usecols=["game_id", "team_id", "is_home", "team_score", "opp_score", "status", "game_type"], low_memory=False)
    sides = sides[(sides["status"] == "final") & (sides["game_type"] == "regular")]

    merged = av.merge(sides[["game_id", "team_id", "is_home"]], on=["game_id", "team_id"])
    home = merged[merged["is_home"] == 1].set_index("game_id")
    away = merged[merged["is_home"] == 0].set_index("game_id")
    both = home.index.intersection(away.index)

    game = pd.DataFrame(index=both)
    game["d_miss_value"] = away.loc[both, "miss_value"].to_numpy() - home.loc[both, "miss_value"].to_numpy()
    game["d_best_out"] = away.loc[both, "best_out"].to_numpy() - home.loc[both, "best_out"].to_numpy()

    tab = pd.read_csv(TABLE, low_memory=False).set_index("game_id")
    h = sides[sides["is_home"] == 1].set_index("game_id")
    tab["home_margin"] = (h["team_score"] - h["opp_score"]).reindex(tab.index)

    df = tab.join(game, how="inner").dropna(subset=["home_margin"])
    for c in SCHEDULE + CONTROLS:
        df[c] = df[c].fillna(0.0)
    y = df["home_margin"].to_numpy(float)

    base, rmse_base = ols(df[CONTROLS + SCHEDULE].to_numpy(float), y, CONTROLS + SCHEDULE)
    ctl_only, _ = ols(df[CONTROLS].to_numpy(float), y, CONTROLS)

    # Published effects: each measured in its OWN specification, because the absence
    # measures are collinear with one another and split unstably when entered together.
    names_v = CONTROLS + SCHEDULE + ["d_miss_value"]
    fit_v, rmse_v = ols(df[names_v].to_numpy(float), y, names_v)
    names_b = CONTROLS + SCHEDULE + ["d_best_out"]
    fit_b, _ = ols(df[names_b].to_numpy(float), y, names_b)

    def cell(frame: pd.DataFrame, key: str) -> dict:
        return {
            "points": round(float(frame.loc[key, "beta"]), 3),
            "t": round(float(frame.loc[key, "t"]), 1),
        }

    facts = {
        "generatedFrom": "hoopR player box scores + games (final regular season only)",
        "sample": {
            "games": int(len(df)),
            "teamGames": int(len(av)),
            "firstSeason": str(df["season"].min()),
            "lastSeason": str(df["season"].max()),
        },
        "definition": {
            "rotationWindowGames": WINDOW,
            "rotationMinutes": ROTATION_MINUTES,
            "minAppearances": MIN_APPEARANCES,
            "qualityGames": QUALITY_GAMES,
        },
        # The headline scale: everything the site measures, in one unit.
        "effectsInPoints": {
            "bestPlayerOut": cell(fit_b, "d_best_out"),
            "homeCourt": {"points": round(float(ctl_only.loc["intercept", "beta"]), 3), "t": round(float(ctl_only.loc["intercept", "t"]), 1)},
            "backToBack": cell(base, "d_is_b2b"),
            "visitingAltitude": cell(base, "d_alt_visit"),
            "priorOvertime": cell(base, "d_prior_ot"),
            "missValuePerPoint": cell(fit_v, "d_miss_value"),
        },
        "frequency": {
            "teamGamesMissingBestPlayer": round(float(av["best_out"].mean()), 4),
            "teamGamesMissingNobody": round(float((av["miss_count"] == 0).mean()), 4),
            "gamesEitherSideMissingBest": round(float((game["d_best_out"].abs() > 0).mean()), 4),
            "meanRotationSize": round(float(av["rotation_size"].mean()), 2),
        },
        # How much of the schedule effect is really just stars sitting. Small, which is the
        # point: it is what stops "your back-to-back result is load management" being true.
        "scheduleHoldsUp": {
            {
                "d_is_b2b": "backToBack",
                "d_alt_visit": "visitingAltitude",
                "d_prior_ot": "priorOvertime",
                "d_density_points": "scheduleDensity",
            }[term]: {
                "scheduleOnly": round(float(base.loc[term, "beta"]), 3),
                "absenceControlled": round(float(fit_v.loc[term, "beta"]), 3),
                "shiftPct": round(
                    100.0 * (float(base.loc[term, "beta"]) - float(fit_v.loc[term, "beta"]))
                    / float(base.loc[term, "beta"]),
                    1,
                ),
            }
            for term in SCHEDULE
        },
        # Published so the page cannot imply it explains games. It does not.
        "noise": {
            "marginStdDev": round(float(np.std(y)), 2),
            "rmseScheduleOnly": round(rmse_base, 3),
            "rmseWithAbsence": round(rmse_v, 3),
        },
        "seasonBestOutRate": [
            {"season": str(s), "rate": round(float(g["best_out"].mean()), 4), "teamGames": int(len(g))}
            for s, g in av.groupby("season", sort=True)
        ],
    }

    JSON_PATH.write_text(json.dumps(facts, indent=2) + "\n")
    log.info("wrote %s", JSON_PATH)


if __name__ == "__main__":
    main()

File: src/lib/availability-facts.ts
/**
 * Every figure the availability surface publishes.
 *
 * These are constants because they describe a finished measurement, not a query: they change
 * only when `ml/availability_facts.py` is re-run. `src/lib/__tests__/availability-facts.test.ts`
 * pins this file against `ml/availability_facts.json`, so a number edited here and nowhere else
 * fails the suite. Same arrangement as `playoff-rest-facts.ts`, for the same reason — the
 * figures on that page were once retyped into three places and drifted.
 *
 * **Everything here is retrospective.** Who sat is known only because the game was played, so
 * none of it forecasts anything. Any copy built on these numbers has to say so: this measures
 * what an absence cost, never who will be available tonight.
 */

/** A measured effect, in points of final margin, with the t-statistic behind it. */
export interface AvailabilityEffect {
  /** Points of margin. Positive means it costs the team carrying it. */
  points: number;
  /** |t| under 2 would mean "not distinguishable from zero". Everything published here clears it. */
  t: number;
}

/** One season's rate of playing without the team's best player. */
export interface BestOutSeason {
  season: string;
  /** 0-1. Share of that season's team-games. */
  rate: number;
  teamGames: number;
}

/** What the measurement was computed over. */
export const AVAILABILITY_SAMPLE = Object.freeze({
  games: 35458,
  teamGames: 70932,
  firstSeason: "1996-97",
  lastSeason: "2025-26",
});

/**
 * How "missing" is defined. Absence is inferred from prior participation rather than from
 * tonight's sheet, because a long-term-injured player may not be listed at all — so the only
 * reliable signal is the rotation the team has actually been using.
 */
export const AVAILABILITY_DEFINITION = Object.freeze({
  rotationWindowGames: 5,
  rotationMinutes: 15.0,
  minAppearances: 2,
  qualityGames: 20,
});

/**
 * The headline: everything this site measures, in one unit.
 *
 * `bestPlayerOut` and `homeCourt` landing within 0.04 points of each other is the finding —
 * losing your best player costs about what playing at home is worth. Each effect comes from
 * its own specification, because the absence measures are collinear with one another and split
 * unstably when entered together.
 *
 * `missValuePerPoint` is the continuous version, used for modelling rather than display: points
 * of margin per point of game score above the team's own replacement level.
 */
export const AVAILABILITY_EFFECTS = Object.freeze({
  bestPlayerOut: Object.freeze({ points: 2.863, t: 17.9 }) as AvailabilityEffect,
  homeCourt: Object.freeze({ points: 2.824, t: 42.3 }) as AvailabilityEffect,
  backToBack: Object.freeze({ points: 1.759, t: 13.4 }) as AvailabilityEffect,
  visitingAltitude: Object.freeze({ points: 1.358, t: 5.1 }) as AvailabilityEffect,
  priorOvertime: Object.freeze({ points: 0.544, t: 2.9 }) as AvailabilityEffect,
  missValuePerPoint: Object.freeze({ points: 0.26, t: 21.9 }) as AvailabilityEffect,
});

/** How often it happens at all. */
export const AVAILABILITY_FREQUENCY = Object.freeze({
  teamGamesMissingBestPlayer: 0.0991,
  teamGamesMissingNobody: 0.4455,
  gamesEitherSideMissingBest: 0.1714,
  meanRotationSize: 8.62,
});

/**
 * The defensive result, and the reason this measurement was worth taking at all.
 *
 * "Your back-to-back effect is really just teams resting their stars" is the obvious objection
 * to the whole rest-advantage premise. Putting absence and the schedule terms in one regression
 * answers it: every schedule term moves by under 8%, so load management explains almost none of
 * what a back-to-back costs.
 */
export const AVAILABILITY_SCHEDULE_HOLDS_UP = Object.freeze({
  backToBack: Object.freeze({ scheduleOnly: 1.759, absenceControlled: 1.641, shiftPct: 6.7 }),
  visitingAltitude: Object.freeze({ scheduleOnly: 1.358, absenceControlled: 1.282, shiftPct: 5.6 }),
  priorOvertime: Object.freeze({ scheduleOnly: 0.544, absenceControlled: 0.501, shiftPct: 7.9 }),
  scheduleDensity: Object.freeze({ scheduleOnly: 0.275, absenceControlled: 0.265, shiftPct: 3.8 }),
});

/**
 * Published so the page cannot imply it explains games. It does not.
 *
 * A margin standard deviation of 13.64 against a residual of 12.44 means everything measured
 * here — team strength included — accounts for a small share of what happens in a basketball
 * game. The effects are real and precisely estimated; they are not a prediction.
 */
export const AVAILABILITY_NOISE = Object.freeze({
  marginStdDev: 13.64,
  rmseScheduleOnly: 12.52,
  rmseWithAbsence: 12.436,
});

/**
 * The load-management era, in one column. Teams play without their best player more than three
 * times as often as they did in 1996-97 — this is context for every other number on the page,
 * since the same absence was a far rarer event across most of the sample.
 */
export const AVAILABILITY_BEST_OUT_BY_SEASON: readonly BestOutSeason[] = Object.freeze([
  Object.freeze({ season: "1996-97", rate: 0.0596, teamGames: 2233 }),
  Object.freeze({ season: "1997-98", rate: 0.0648, teamGames: 2378 }),
  Object.freeze({ season: "1998-99", rate: 0.0641, teamGames: 1450 }),
  Object.freeze({ season: "1999-00", rate: 0.0509, teamGames: 2378 }),
  Object.freeze({ season: "2000-01", rate: 0.0673, teamGames: 2378 }),
  Object.freeze({ season: "2001-02", rate: 0.0656, teamGames: 2378 }),
  Object.freeze({ season: "2002-03", rate: 0.0459, teamGames: 2373 }),
  Object.freeze({ season: "2003-04", rate: 0.077, teamGames: 2378 }),
  Object.freeze({ season: "2004-05", rate: 0.0772, teamGames: 2460 }),
  Object.freeze({ season: "2005-06", rate: 0.0675, teamGames: 2460 }),
  Object.freeze({ season: "2006-07", rate: 0.0809, teamGames: 2460 }),
  Object.freeze({ season: "2007-08", rate: 0.0707, teamGames: 2460 }),
  Object.freeze({ season: "2008-09", rate: 0.0829, teamGames: 2460 }),
  Object.freeze({ season: "2009-10", rate: 0.078, teamGames: 2460 }),
  Object.freeze({ season: "2010-11", rate: 0.0789, teamGames: 2460 }),
  Object.freeze({ season: "2011-12", rate: 0.1136, teamGames: 1980 }),
  Object.freeze({ season: "2012-13", rate: 0.0964, teamGames: 2458 }),
  Object.freeze({ season: "2013-14", rate: 0.0935, teamGames: 2460 }),
  Object.freeze({ season: "2014-15", rate: 0.1016, teamGames: 2460 }),
  Object.freeze({ season: "2015-16", rate: 0.0829, teamGames: 2460 }),
  Object.freeze({ season: "2016-17", rate: 0.0817, teamGames: 2460 }),
  Object.freeze({ season: "2017-18", rate: 0.115, teamGames: 2460 }),
  Object.freeze({ season: "2018-19", rate: 0.1106, teamGames: 2460 }),
  Object.freeze({ season: "2019-20", rate: 0.1308, teamGames: 2118 }),
  Object.freeze({ season: "2020-21", rate: 0.1417, teamGames: 2160 }),
  Object.freeze({ season: "2021-22", rate: 0.1671, teamGames: 2460 }),
  Object.freeze({ season: "2022-23", rate: 0.1772, teamGames: 2460 }),
  Object.freeze({ season: "2023-24", rate: 0.1293, teamGames: 2460 }),
  Object.freeze({ season: "2024-25", rate: 0.1918, teamGames: 2450 }),
  Object.freeze({ season: "2025-26", rate: 0.1947, teamGames: 2460 }),
]);
